lib/Fmt_ast.ml

```diff
@@ -2768,14 +2810,41 @@ and fmt_expression c ?(box = true) ?(pro = noop) ?eol ?parens
       in
       let outer_wrap = has_attr && parens in
       let inner_wrap = has_attr || parens in
+      let fmt_lt_exp_element (lbl, exp) =
+        let exp = sub_exp ~ctx exp in
+        match lbl with
+        | None -> fmt_expression c exp
+        | Some lbl ->
+            let punned =
+              match exp.ast.pexp_desc with
+              | Pexp_ident {txt= Lident var; _} ->
+                  String.equal lbl.txt var
+                  && List.is_empty exp.ast.pexp_attributes
+              | _ -> false
+            in
+            let punned_with_constraint =
+              match exp.ast.pexp_desc with
+              | Pexp_constraint
+                  ({pexp_desc= Pexp_ident {txt= Lident var; _}; _}, _) ->
+                  String.equal var lbl.txt
+                  && List.is_empty exp.ast.pexp_attributes
+              | _ -> false
+            in
+            if punned then
+              Cmts.fmt c lbl.loc
+              @@ Cmts.fmt c exp.ast.pexp_loc
+              @@ hovbox 0 (str "~" $ str lbl.txt)
+            else if punned_with_constraint then
+              Cmts.fmt c lbl.loc @@ str "~" $ fmt_expression c exp
```

Comment:
Related to the consistency comment. Making the following change to both seem to produce similar results:

```suggestion
              Cmts.fmt c lbl.loc @@ (str "~" $ fmt_expression c exp)
```
gives:
```diff
from
let _ = ~z, ~(y(* baz *) : int)
let ~z:42, ~(y(* baz *) : int) = ()

to
-let _ = ~z, ~(* baz *) (y : int)
+let _ = ~z, ~(y : int) (* baz *)

-let ~z:42, ~((* baz *) y : int) = ()
+let ~z:42, ~(y : int) (* baz *) = ()
```